### ISSUE TITLE
Generate modern sbom formats when releasing etcd

### ISSUE THREAD

**Issue by idunbarh** [etcd-io/etcd#18902]
### What would you like to be added?

I'd like to contribute SBOM generation to the release process of this project in both cyclonedx and spdx formats.

I'm part of https://github.com/CISA-SBOM-Community/SBOM-Generation thats building reference implementations for "good" SBOM generation and we thought etcd would be a great candidate.

### Why is this needed?

SBOMs are becoming a common part of software releases because they provide insight into what dependencies are used in a project. This allows better vulnerability management.

**Comment by jmhbnz** [comment]
Hi @idunbarh many thanks for raising this and volunteering. For background we last touched the SBOM process for etcd in https://github.com/etcd-io/etcd/issues/15747 and the process essentially relies on a very antiquated approach to generate the json file https://github.com/etcd-io/etcd/blob/main/bill-of-materials.json.

Adopting new cyclonedx and spdx format support is definitely something I would support.  We would need to still retain the old json format sbom process in the stable release branches until the next minor release at least in order to not break any existing processes but I am very keen to forge ahead implementing this in `main` initially.

cc @ivanvc, @ahrtr, @serathius  for any further thoughts on this.

/assign @idunbarh

**Comment by jmhbnz** [comment]
/retitle Generation modern sbom formats when releasing etcd

**Comment by ivanvc** [comment]
I think this is a great idea. I spoke with @puerco (sig release) regarding this at KubeCon.

**Comment by ahrtr** [comment]
Thanks for raising this discussion. The json format of [SPDX](https://spdx.dev/) seems a better choice to me. The only minor concern is that [SBOM-Generation](https://github.com/CISA-SBOM-Community/SBOM-Generation) is implemented with Python. It means that we have to get python installed in contributors' dev environment to update SBOM.

Alternatives:
- https://github.com/anchore/syft
- generate a tool based on https://github.com/spdx/tools-golang

**Comment by ahrtr** [comment]
> I'll use syft for this implementation.

Do you mean https://github.com/anchore/syft? If yes, then looks good to me.

**Comment by jmhbnz** [comment]
Discussed during sig-etcd triage meeting. Checking in, @idunbarh are you still keen to work on this?

**Comment by jmhbnz** [comment]
Discussed during sig-etcd triage meeting, opening this up for other contributors given the time passed and adding help wanted.

**Comment by vivekpatani** [comment]
@jmhbnz can be assigned to me.

**Comment by jmhbnz** [comment]
Discussed during sig-etcd triage meeting - Hey @vivekpatani gentle ping are you still working on this one?

**Comment by vivekpatani** [comment]
@jmhbnz sorry, took a long break, will try to get to it this, week. Sorry about that.

**Comment by vivekpatani** [comment]
## Recommendation

I'm with @idunbarh on this for using [Anchore Syft](https://github.com/anchore/syft) - Go-based, supports both SPDX and CycloneDX, no Python dependency

## Plan

### Phase 1: Add modern SBOM generation alongside existing

- Create new script `scripts/generate-modern-sbom.sh`
- Generate both `SPDX` and `CycloneDX` formats
- Keep existing [bill-of-materials.json](bill-of-materials.json:1) process intact

### Phase 2: Integrate into release process

- Modify `scripts/release.sh` to include SBOM generation after build step
- Upload SBOM files to `gs://etcd` alongside other artifacts
- Include SBOM files in GitHub releases

### Phase 3: Testing integration

- Add SBOM validation to `scripts/test.sh`
- Ensure SBOM files are generated correctly and contain expected dependencies

## Projected File Structure

```bash
├── bill-of-materials.json          # Keep existing (legacy)
├── bill-of-materials.override.json # Keep existing
├── sbom.spdx.json                 # New SPDX format
├── sbom.cyclonedx.json            # New CycloneDX format
└── scripts/
    ├── generate-modern-sbom.sh    # New script
    └── updatebom.sh               # Keep existing
```

## Script

```bash
#!/usr/bin/env bash

set -euo pipefail

VERSION=${1:-"dev"}
OUTPUT_DIR=${2:-"."}

# Install syft if not present
if ! command -v syft >/dev/null; then
    echo "Installing syft..."
    curl -sSfL https://raw.githubusercontent.com/anchore/syft/main/install.sh | sh -s -- -b /usr/local/bin
fi

# Generate SPDX format
syft packages . -o spdx-json="${OUTPUT_DIR}/sbom.spdx.json"

# Generate CycloneDX format  
syft packages . -o cyclonedx-json="${OUTPUT_DIR}/sbom.cyclonedx.json"

echo "SBOM files generated successfully"
```

Feedback would be appreciated: @jmhbnz @ahrtr @ivanvc

**Comment by ivanvc** [comment]
Running both side to side sounds like a good migration plan. I would suggest putting a POC PR, so we can see this in action. Thanks, @vivekpatani :)

**Comment by idunbarh** [comment]
@vivekpatani Just a couple of suggestions.

CycloneDX standard recommends using `blah.cdx.json` instead of `blah.cyclonedx.json`.

Take a look at tools like [sbomasm](https://github.com/interlynk-io/sbomasm) to edit information like supplier, copyright, and other metadata thats not populated by syft. We have some example github workflows we put together for the [OpenSSF community here](https://github.com/SBOM-Community/SBOM-Generation/blob/main/.github/workflows/phase_1_keycloak.yml#L84).

Sorry I dropped the ball on the issue, and @vivekpatani thanks for jumping in! I'm got some more time to help with this. @vivekpatani let me know if you want any support.

**Comment by jmhbnz** [comment]
Sound like a great plan @vivekpatani - thanks for getting this back underway. Please ping me in the review for the phase one script pr 🙏🏻

**Comment by vivekpatani** [comment]
@jmhbnz @ivanvc @idunbarh @ahrtr could use your feedback to improve the PR: https://github.com/etcd-io/etcd/pull/20303, thank you.

**Comment by vivekpatani** [comment]
Hi team, @ivanvc does this still need work? Seems like this was referenced in a couple of more issues, I have cycles this sprint to work on this.

**Comment by ivanvc** [comment]
@vivekpatani, yes, please. I believe the PR #20303 stalled. It needs to address some comments to move forward.

**Comment by tushar1977** [comment]
Hi, can you assign this issue to me?I'd love to work on this

**Comment by vivekpatani** [comment]
Sorry for the radio silence, I've been making a bit of progress since the last week, @tushar1977 I'll try to submit whatever I have in case if it isn't complete by Friday if that's okay? Once again sorry for the slow progress.

**Comment by ahrtr** [comment]
I think we need to resolve this.

The existing BOM generation script/tool has a flaw. It essentially collects dependency packages by executing a command similar to the one below.

The problem is that the result is platform-dependent. Some Go source files are guarded by build tags, so they may be included or excluded depending on the target platform. As a result, different platforms may produce different dependency sets, making the BOM unstable and non-reproducible.

```
go list -f '{{range .Deps}}{{.}}|{{end}}' ./... ./api/... ./cache/... ./client/pkg/... ./client/v3/... ./etcdctl/... ./etcdutl/... ./pkg/... ./server/... ./tests/... 
```

One example is the go source files (e.g. [example](https://github.com/etcd-io/etcd/blob/main/tests/antithesis/test-template/robustness/finally/main.go#L15) ) which use the `github.com/antithesishq/antithesis-sdk-go` have build tag `//go:build cgo && amd64`, so they aren't scanned at all on my Mac M1(arm64), so it always fails when I execute `make verify-bom`

```
$ make verify-bom
PASSES="bom" ./scripts/test.sh
Running with --race
Starting at: Thu 30 Apr 2026 16:04:15 BST

'bom' started at Thu 30 Apr 2026 16:04:15 BST
Checking bill of materials...
% 'cp' 'go.sum' 'go.sum.tmp'
% 'cp' 'go.mod' 'go.mod.tmp'
% (cd tools/mod && 'go' 'install' 'github.com/appscodelabs/license-bill-of-materials')
% '/Users/wachao/go/bin/license-bill-of-materials' '--override-file' './bill-of-materials.override.json' './...' './api/...' './cache/...' './client/pkg/...' './client/v3/...' './etcdctl/...' './etcdutl/...' './pkg/...' './server/...' './tests/...'
% 'cp' 'go.sum.tmp' 'go.sum'
% 'cp' 'go.mod.tmp' 'go.mod'
21,29d20
< 		"project": "github.com/antithesishq/antithesis-sdk-go",
< 		"licenses": [
< 			{
< 				"type": "MIT License",
< 				"confidence": 1
< 			}
< 		]
< 	},
< 	{
modularized licenses do not match given bill of materials
FAIL: 'bom' FAILED at Thu 30 Apr 2026 16:04:25 BST
make: *** [verify-bom] Error 255
```

@vivekpatani are you still working on this per your plan https://github.com/etcd-io/etcd/issues/18902#issuecomment-2978143716?

**Comment by vivekpatani** [comment]
Hadn't looked at it in a bit. But I can take a look at it tomorrow, will fix up the PR, and re tag Ivan and you.
